pkg/suggestion/genetic_algorithm/gene.go

```diff
@@ -0,0 +1,144 @@
+package genetic_algorithm
+
+// for gene with continuous variables in float64s
+type ContinuousGeneType struct {
+	min float64
+	max float64
+}
+
+func NewContinuousGeneType(min, max float64) *ContinuousGeneType {
+	return &ContinuousGeneType{min, max}
+}
+
+func (cg *ContinuousGeneType) generateRandom() float64 {
+	return getContinuousRandom(cg.min, cg.max)
+}
+
+func (cg *ContinuousGeneType) setGeneRandomValue(g *Gene) {
+	g.setGeneValue(cg.generateRandom())
+}
+
+// for gene with discrete variables in int
+type DiscreteGeneType struct {
+	min int
+	max int
+}
+
+func NewDiscreteGeneType(min, max int) *DiscreteGeneType {
+	return &DiscreteGeneType{min, max}
+}
+
+func (dg *DiscreteGeneType) generateRandom() int {
+	return getDiscreteRandom(dg.min, dg.max)
+}
+
+func (dg *DiscreteGeneType) setGeneRandomValue(g *Gene) {
+	g.setGeneValue(dg.generateRandom())
+}
+
+// for gene with categorical variables in string
+type CategoricalGeneType struct {
+	category []string
+}
+
+func NewCategoricalGeneType(category []string) *CategoricalGeneType {
+	return &CategoricalGeneType{category}
+}
+
+func (cg *CategoricalGeneType) generateRandom() int {
+	return getCategoricalRandom(cg.category)
+}
+
+func (cg *CategoricalGeneType) setGeneRandomValue(g *Gene) {
+	g.setGeneValue(cg.category[cg.generateRandom()])
+}
+
+type setGeneRandomValueInterface interface {
+	setGeneRandomValue(g *Gene)
+}
+
+// map for gene name and gene type
+type GeneTypeMap map[string]interface{} // map[geneName]geneType
+
+func NewGeneTypeMap(geneNames []string, geneTypes []interface{}) GeneTypeMap {
+	geneTypeMap := make(map[string]interface{})
+	for i := 0; i < len(geneNames); i++ {
+		geneTypeMap[geneNames[i]] = geneTypes[i]
+	}
+	return geneTypeMap
+}
+
+// gene struct; one name and one value per one gene
+type Gene struct {
+	geneName  string      // name of the gene
+	geneValue interface{} // value of the gene; must fit with the gene type
+}
+
+func NewGene(geneName string) *Gene {
+	return &Gene{geneName, ""}
+}
+
+func (g *Gene) setGeneValue(v interface{}) {
+	g.geneValue = v
+}
+
+func (g *Gene) setRandomGeneValue(s setGeneRandomValueInterface) {
+	s.setGeneRandomValue(g)
+}
+
+// set random value depending on the gene type
+func setRandomGeneValueWithType(gene *Gene, geneTypeMap GeneTypeMap) {
+	switch s := geneTypeMap[gene.geneName].(type) {
+	case CategoricalGeneType:
```

Comment:
Why you use Type Switch?
Why not make an interface for GeneType and only call its func.

Reply:
I have the interface for the geneType on lines 56 to 58 as like below.
 
type setGeneRandomValueInterface interface {
	setGeneRandomValue(g *Gene)
}